### PR TITLE
skip evaluating Pods during requeueing when they're scheduling-gated

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

A new scheduler_perf test `SchedulingGatedPodsWithPodAffinityImpactForThroughput` (https://github.com/kubernetes/kubernetes/pull/125550) highlights the degradation when there are many gated Pods with PodAffinity.

The degradation could happen when all these conditions are met:

- There are many gated Pods with PodAffinity.
- There are many Pods getting scheduled, that is, the scheduling queue keeps receiving AssignedPodAdd cluster events.

`getUnschedulablePodsWithMatchingAffinityTerm`, which is called every time the queue receives `AssignedPodAdd`, visits all unschedulable Pods and checks if PodAffinity of each unscheduled Pod matches with a newly scheduled Pod.

So, it's fundamentally the same cause with https://github.com/kubernetes/kubernetes/issues/124384; `getUnschedulablePodsWithMatchingAffinityTerm` slows down, causes the queue taking a lock for a long time, and consequently the scheduling throughput is impacted negatively.

---

`SchedulingGatedPodsWithPodAffinityImpactForThroughput` shows this PR bring back the throughput above 300 pods/s:

- Before this PR: SchedulingThroughput: run#1 171 pods/s, run#2 172 pods/s
- After this PR: SchedulingThroughput: run#1 419 pods/s, run#2 370 pods/s

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

I'm not sure if it deserves a cherry-pick since the scenario is more rare than https://github.com/kubernetes/kubernetes/issues/124384.
Defer to @kubernetes/sig-scheduling-leads for the decision.

#### Does this PR introduce a user-facing change?

```release-note
Fix a degradation that happens when there are many scheduling-gated Pods with PodAffinity.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/hold

To make sure it goes through an approval from other approvers.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -1234,12 +1235,24 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
...
+// We know it depends on the logic of InterPodAffinity plugin, which is not ideal since the scheduling queue should be pure,
+// and after all QueueingHint work finishes, we'll cleanup this function along with PreEnqueueCheck.
+//
 // NOTE: this function assumes lock has been acquired in caller.
 func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(logger klog.Logger, pod *v1.Pod) []*framework.QueuedPodInfo {
 	nsLabels := interpodaffinity.GetNamespaceLabelsSnapshot(logger, pod.Namespace, p.nsLister)
 
 	var podsToMove []*framework.QueuedPodInfo
 	for _, pInfo := range p.unschedulablePods.podInfoMap {
+		if pInfo.Gated && pInfo.UnschedulablePlugins.Has(names.SchedulingGates) {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:1250]: It takes the same approach as https://github.com/kubernetes/kubernetes/pull/125527, using `pInfo.UnschedulablePlugins.Has` to make this shortcut available only for the scheduling gate plugin.

**Comment by alculquicondor** [comment]
Is this a different approach to solve #125223?

**Comment by sanposhiho** [comment]
@alculquicondor No, unfortunately, this is for a separate, different issue.

**Comment by alculquicondor** [comment]
I think we should not merge or cherry-pick this and focus on the long term fix.

**Comment by sanposhiho** [comment]
Agree, if we rearchitect activeQ, this won't impact the scheduling throughput negatively, and also I guess we can say this scenario is fairly rare, which isn't urgent to fix, either way.

/close

Feel free to reopen it when find it useful in some way.

**Pull request closed without merging**